File: test/conformance/rust/rows/C8.rs
// C8 — fixed-point F-shift / bits(N) (docs/FIXED-FORM-ALGORITHM.md:80-84, 351-352)
//
// THE LAW (inlined from the spec):
//   A fixed-point field's bounds are in VALUE UNITS and its storage is raw,
//   so both ends are shifted by F first; a bits(N) clamps to 2^N - 1.
//
// This test reaches the generated Rust tables through the same fixed-form
// load path the driver and the rust-fixedform test use.

extern crate tabledemo;
extern crate benchfixed;

fn check(ok: bool, msg: &str) {
    if !ok {
        panic!("FAIL: {msg}");
    }
    println!("  ok: {msg}");
}

#[test]
fn c8_fixed_point_f_shift_and_bits_n() {
    // ================================================================
    // PART 1: bits(N) clamps to 2^N - 1
    // ================================================================
    // RangedWidths has bits(12) at body offset 20. bits(12) declares
    // range [0, 4095] = [0, 2^12 - 1]. A forged value of 4096 must
    // land at 4095 after the bounds pass.

    {
        let mut one = tabledemo::RangedWidthsRow::default();
        one.b12 = 3000;
        one.b48 = 0x1234_5678_9ABC_u64;
        let values = [one];
        let mut file = vec![0u8; tabledemo::ranged_widths_fixed_measure(1)];
        check(
            tabledemo::ranged_widths_fixed_save(&values, &mut file) == Some(file.len()),
            "bits(N): the record writes",
        );

        let mut back = [tabledemo::RangedWidthsRow::default(); 2];
        let mut plan = [tabledemo::TableFixedEntry::default(); 512];
        let mut remap = [0u16; 512];
        let mut report = tabledemo::TableFixedReport::default();
        let n = tabledemo::ranged_widths_fixed_load(&mut back, &file, &mut plan, &mut remap, &mut report);
        check(n == Some(1), "bits(N): the clean record reads");
        check(
            back[0].b12 == 3000 && report.clamped == 0,
            "bits(N): value inside bits(12) lands untouched",
        );

        // File layout: header(16) + layout_len(4) + layout(123) + record_hash(8) + body
        // Body offset = 16 + 4 + 123 + 8 = 151
        let body = tabledemo::TABLE_FIXED_HEADER_BYTES + 4
            + tabledemo::RANGED_WIDTHS_FIXED_BLOCK.len() + 8;
        // b12 sits at body offset 20 (b8=0, b16=4, b32=8, b64=12)
        let mut forged = file.clone();
        forged[body + 20] = 0x00;
        forged[body + 21] = 0x10; // 4096 = 0x1000 LE
        forged[body + 22] = 0x00;
        forged[body + 23] = 0x00;
        let mut report = tabledemo::TableFixedReport::default();
        let n = tabledemo::ranged_widths_fixed_load(&mut back, &forged, &mut plan, &mut remap, &mut report);
        check(n == Some(1), "bits(N): the forged record still reads");
        check(back[0].b12 == 4095, "bits(N): 4096 -> 4095 = 2^12 - 1");
        check(report.clamped == 1, "bits(N): counts ONE clamped");

        // bits(48) at body offset 24: u64::MAX past 2^48-1
        let mut forged = file.clone();
        for i in 0..8 { forged[body + 24 + i] = 0xFF; }
        let mut report = tabledemo::TableFixedReport::default();
        tabledemo::ranged_widths_fixed_load(&mut back, &forged, &mut plan, &mut remap, &mut report);
        check(
            back[0].b48 == 0xFFFF_FFFF_FFFF && report.clamped == 1,
            "bits(N): bits(48) past 2^48-1 clamps to 2^48-1",
        );

        // bits(32) fills its lane exactly — no clamp fires
        let mut forged = file.clone();
        for i in 0..4 { forged[body + 8 + i] = 0xFF; }
        let mut report = tabledemo::TableFixedReport::default();
        tabledemo::ranged_widths_fixed_load(&mut back, &forged, &mut plan, &mut remap, &mut report);
        check(
            back[0].b32 == 0xFFFF_FFFF && report.clamped == 0,
            "bits(N): bits(32) filling its lane clamps nothing (negative control)",
        );
    }

    // ================================================================
    // PART 2: fixed-point F-shift
    // ================================================================
    // BenchMixedRow.server_time is fixed(24, 8) | min = 0, max = 65535.
    // F = 8, so raw bounds are [0, 65535 << 8] = [0, 16776960].
    //
    // The law says "both ends are shifted by F first". To prove it:
    //   - raw value 65536 is past the value-unit max (65535) but INSIDE
    //     the shifted max (16776960) → must NOT clamp
    //   - raw value 16776961 is past the shifted max → clamps to 16776960

    {
        // Directly test the clamp_body function on a BenchMixedRow.
        // This is the function that implements the F-shift law.

        // Test 1: value inside shifted bounds, past value-unit max
        let mut value = benchfixed::BenchMixedRow::default();
        value.server_time = 65536; // past value-unit max 65535, inside shifted max 16776960
        value.ping = 100;
        let mut clamped: i32 = 0;
        let mut damaged: i32 = 0;
        benchfixed::bench_mixed_fixed_clamp_body(&mut value, &mut clamped, &mut damaged);
        check(
            value.server_time == 65536 && clamped == 0,
            "F-shift: raw 65536 (past value-unit max) survives — bound is SHIFTED by F=8",
        );

        // Test 2: value past shifted max
        let mut value = benchfixed::BenchMixedRow::default();
        value.server_time = 16776961; // one past shifted max
        value.ping = 100;
        let mut clamped: i32 = 0;
        let mut damaged: i32 = 0;
        benchfixed::bench_mixed_fixed_clamp_body(&mut value, &mut clamped, &mut damaged);
        check(
            value.server_time == 16776960,
            "F-shift: 16776961 -> 16776960 = 65535 << 8 (the SHIFTED max)",
        );
        check(clamped == 1, "F-shift: counts ONE clamped");

        // Test 3: value at shifted max — clean
        let mut value = benchfixed::BenchMixedRow::default();
        value.server_time = 16776960; // exactly at shifted max
        value.ping = 100;
        let mut clamped: i32 = 0;
        let mut damaged: i32 = 0;
        benchfixed::bench_mixed_fixed_clamp_body(&mut value, &mut clamped, &mut damaged);
        check(
            value.server_time == 16776960 && clamped == 0,
            "F-shift: raw 16776960 (exactly the shifted max) clamps nothing",
        );

        // Test 4: ping is ufixed(8,8) | min=0, max=250, F=8
        // Raw bounds: [0, 250 << 8] = [0, 64000]
        let mut value = benchfixed::BenchMixedRow::default();
        value.server_time = 0;
        value.ping = 64001; // one past the shifted max
        let mut clamped: i32 = 0;
        let mut damaged: i32 = 0;
        benchfixed::bench_mixed_fixed_clamp_body(&mut value, &mut clamped, &mut damaged);
        check(
            value.ping == 64000,
            "F-shift: ufixed(8,8) ping: 64001 -> 64000 = 250 << 8",
        );
        check(clamped == 1, "F-shift: ping counts ONE clamped");
    }

    println!("C8: all assertions passed");
}

File: test/conformance/dart/rows/C8.dart
// test/conformance/dart/rows/C8.dart — the dart/C8 cell, "fixed-point
// F-shift / bits(N)" (docs/roadmap.sexp at main eb12ceb3).
//
// THE LAW (docs/FIXED-FORM-ALGORITHM.md:81, §1.2's LEAF row at line 110,
// §3's C table at line 188):
//
//   bits(N): N in 1..8 -> kind 6 (u8), N in 9..16 -> kind 7 (u16),
//            N in 17..32 -> kind 8 (u32), N in 33..64 -> kind 9 (u64);
//            the declared STORAGE WIDTH is 4 for N<=32 and 8 for N>=33;
//            the read DECLARES a WIDTH CLAMP of 2^N - 1, which the storage
//            type holds exactly when N IS the storage width, so the clamp
//            is ELIDED for N=32 (storage u32 is also 2^32 - 1) and N=64
//            (storage u64 is also 2^64 - 1).
//
//   fixed(I,F) F-shift: a fixed-point value rides raw at its storage width
//            on this form: the storage is the scaled integer and NOTHING
//            HERE DIVIDES BY 2^F. The bound COMPARED AGAINST IS THE
//            DECLARED WHOLE-UNIT BOUND SHIFTED BY F ONTO THAT SAME RAW
//            SCALE (§3, docs/FIXED-FORM-ALGORITHM.md:866). The compare is
//            `bound * 2^F` against the raw integer.
//
// PRODUCTION PATH this test drives (every stage generated):
//   rangedWidthsFixedSave       the write (internal/codegen/darttable/fixeddart.go, rangedWidthsFixedWriteBody)
//   rangedWidthsFixedDecode     the read every share, including the WIDTH CLAMP for bits(N) (TabledemoFixed.dart, rangedWidthsFixedDecode)
//   rangedWidthsFixedLoad       the load, identity plan
//   TableFixedLayout.{kind,size} the parser used here to verify the layout's own (kind, size) for each bits(N) entry
//
// WHY THIS TEST EXISTS — the cell title joins TWO clauses the dart leg
// must satisfy. The dart leg has ONE fixed-form TYPE THAT COVERS bits(N):
// tables/examples/Ranges.schema pins RangedWidths with six bits(N) widths
// (b8, b12, b16, b32, b48, b64). The dart leg HAS NO SCHEMA WITH fixed(I,F):
// a search over tables/examples, tables/pointers, tables/block and
// tables/blockhome found no `fixed(` declaration. The bits(N) clause is
// what is asserted below; the fixed(I,F) F-shift CLAUSE IS THE LEG'S
// UNPROVED part — named in RESULT.md and as a follow-up.
//
// THE VECTOR is a single in-range record written by this build's own
// rangedWidthsFixedSave at the field's exact bound ($maskedAt), read back
// through the identity plan: any bits(N) value ridden at the declared
// storage width round-trips byte-identical. The control is one byte past
// the WIDTH CLAMP — a u32 pattern wider than 2^N-1 lands the clamp at the
// law-named value with exactly ONE clamp counter and no other counter
// touched.
//
// Run from the repo root: dart run test/conformance/dart/rows/C8.dart
// Exit 0 green / exit 1 red, one line per assertion.

import 'dart:io';
import 'dart:typed_data';

import '../../../../build/tables-generated-dart/examples/RangesFixed.dart';
import '../../../../build/tables-generated-dart/examples/TabledemoFixed.dart'
    show
        RangedWidths,
        TableFixedLayout,
        TableFixedPlan,
        TableFixedRefusal,
        TableFixedReport;

int failures = 0;

void check(String what, bool ok) {
  print('${ok ? 'ok' : 'FAIL'} - $what');
  if (!ok) {
    failures++;
  }
}

// fnv1a64 over a wire name's ASCII bytes, identical to the layout's own
// hash. Used to map the layouts' id values back to the field name so the
// (kind, size) check is against the NAME, not against an index that could
// move on a future field reorder.
int fnv1a64(String name) {
  final bytes = name.codeUnits;
  var h = 0xcbf29ce484222325;
  for (final b in bytes) {
    h = (h ^ b) & 0xffffffffffffffff;
    h = (h * 0x100000001b3) & 0xffffffffffffffff;
  }
  return h;
}

// THE EXPECTED (kind, size, maxValue at 2^N-1) for each bits(N) field
// RangedWidths carries. The kind is the unsigned integer kind the bit
// count picks; size is the declared storage width (4 for N<=32, 8 for
// N>=33); maxValue is the law's clamp target.
class BitsCase {
  final int kind;
  final int size;
  final int maxValue;
  final int maskLow64;
  const BitsCase(this.kind, this.size, this.maxValue, this.maskLow64);
}

const Map<String, BitsCase> _expectations = {
  'b8':  BitsCase(6, 4, 0xFF,        0x00000000000000FF),
  'b12': BitsCase(7, 4, 0xFFF,       0x0000000000000FFF),
  'b16': BitsCase(7, 4, 0xFFFF,     0x000000000000FFFF),
  'b32': BitsCase(8, 4, 0xFFFFFFFF, 0x00000000FFFFFFFF),
  'b48': BitsCase(9, 8, -1 & 0xFFFFFFFFFFFF, 0x0000FFFFFFFFFFFF), // 2^48 - 1
  'b64': BitsCase(9, 8, -1,         0xFFFFFFFFFFFFFFFF), // 2^64 - 1
};

// LAYOUT-LEVEL ASSERTIONS: the COMPILER PINS (kind, size) PER ENTRY at
// build time against the layout this build ships. A read through the
// PUBLIC TableFixedLayout.{id,kind,size} accessor is the single source
// for "is the law being implemented by THIS BUILD'S compilation?" —
// anything else reads WHAT THE WRITER WROTE, not what the layout itself
// declares, and would let a writer/store mismatch pass.
void assertLayout(TableFixedLayout layout) {
  if (layout.count != _expectations.length + 1) {
    print(
      'C8 layout: expected '
      '${_expectations.length + 1} entries (RangedWidths + 6 bits), '
      'got ${layout.count}',
    );
    failures++;
    return;
  }
  for (var i = 1; i < layout.count; i++) {
    final id = layout.id(i);
    String? name;
    for (final n in _expectations.keys) {
      if (fnv1a64(n) == id) {
        name = n;
        break;
      }
    }
    if (name == null) {
      continue;
    }
    final exp = _expectations[name]!;
    check(
      'C8 layout $name: kind = ${exp.kind} (got ${layout.kind(i)})',
      layout.kind(i) == exp.kind,
    );
    check(
      'C8 layout $name: size = ${exp.size} (got ${layout.size(i)})',
      layout.size(i) == exp.size,
    );
  }
}

// A CLEAN READ through the identity plan: each bit field set to its u64
// IMAGE of the in-range mask, no clamp, round trip preserves the value.
(RangedWidths, TableFixedReport) readClean(
  int b8,
  int b12,
  int b16,
  int b32,
  int b64,
  int b48,
) {
  final value = RangedWidths()
    ..b8 = b8
    ..b12 = b12
    ..b16 = b16
    ..b32 = b32
    ..b64 = b64
    ..b48 = b48;
  final bytes = Uint8List(rangedWidthsFixedMeasure(1));
  rangedWidthsFixedSave([value], 1, bytes);
  final out = RangedWidths();
  final report = TableFixedReport();
  final n = rangedWidthsFixedLoad(
    [out],
    1,
    bytes,
    bytes.length,
    rangedWidthsFixedNewPlan(),
    report,
  );
  if (n != 1) {
    failures++;
    print('FAIL C8 clean read returned $n records (refused=${report.refused})');
  }
  return (out, report);
}

// Round-trip EACH bits(N) AT its in-range bound; clamped == 0, value reads
// back identically.
void assertRoundTrip() {
  final cases = <(String, int)>[
    ('b8',  _expectations['b8']!.maxValue),
    ('b12', _expectations['b12']!.maxValue),
    ('b16', _expectations['b16']!.maxValue),
    ('b32', _expectations['b32']!.maxValue),
    ('b48', _expectations['b48']!.maxValue),
    ('b64', _expectations['b64']!.maxValue & 0x7FFFFFFFFFFFFFFF),
  ];
  for (final c in cases) {
    int b8 = _expectations['b8']!.maxValue;
    int b12 = _expectations['b12']!.maxValue;
    int b16 = _expectations['b16']!.maxValue;
    int b32 = _expectations['b32']!.maxValue;
    int b48 = _expectations['b48']!.maxValue;
    int b64 = _expectations['b64']!.maxValue & 0x7FFFFFFFFFFFFFFF;
    if (c.$1 == 'b8') b8 = c.$2;
    if (c.$1 == 'b12') b12 = c.$2;
    if (c.$1 == 'b16') b16 = c.$2;
    if (c.$1 == 'b32') b32 = c.$2;
    if (c.$1 == 'b48') b48 = c.$2;
    if (c.$1 == 'b64') b64 = c.$2;
    final (back, report) = readClean(b8, b12, b16, b32, b64, b48);
    check(
      'C8 round trip: ${c.$1} at mask=0x${c.$2.toRadixString(16)}: '
      'clamped == 0 (got ${report.clamped})',
      report.clamped == 0,
    );
    final got = switch (c.$1) {
      'b8'  => back.b8,
      'b12' => back.b12,
      'b16' => back.b16,
      'b32' => back.b32,
      'b48' => back.b48,
      'b64' => back.b64,
      _ => 0,
    };
    final exp = c.$2;
    final ok = got == exp;
    check(
      'C8 round trip: ${c.$1} read back = 0x${exp.toRadixString(16)} '
      '(got 0x${got.toRadixString(16)})',
      ok,
    );
  }
}

// FORGED READ: one bit field set one PAST the declared bound. The WIDTH
// CLAMP lands at 2^N - 1 exactly, with exactly ONE clamp and nothing
// else moved. A wrong mask (e.g., 2^N instead of 2^N - 1) turns this
// test RED on the offending field; a wrong storage width (a layout
// admitting size 1 for kind 6 instead of size 4) fails at LAYOUT-LEVEL
// ASSERTION above.
void assertClamp() {
  void one(String field, int rawValue, int clampedAt) {
    final b8  = field == 'b8'  ? rawValue : _expectations['b8']!.maxValue;
    final b12 = field == 'b12' ? rawValue : _expectations['b12']!.maxValue;
    final b16 = field == 'b16' ? rawValue : _expectations['b16']!.maxValue;
    final b32 = field == 'b32' ? rawValue : _expectations['b32']!.maxValue;
    final b64 = field == 'b64' ? rawValue : _expectations['b64']!.maxValue & 0x7FFFFFFFFFFFFFFF;
    final b48 = field == 'b48' ? rawValue : _expectations['b48']!.maxValue;
    final (back, report) = readClean(b8, b12, b16, b32, b64, b48);
    check(
      'C8 clamp $field raw=0x${rawValue.toRadixString(16)}: '
      'clamped == 1 (got ${report.clamped})',
      report.clamped == 1,
    );
    final got = switch (field) {
      'b8'  => back.b8,
      'b12' => back.b12,
      'b16' => back.b16,
      'b32' => back.b32,
      'b48' => back.b48,
      'b64' => back.b64,
      _ => 0,
    };
    check(
      'C8 clamp $field raw=0x${rawValue.toRadixString(16)}: '
      'lands at 0x${clampedAt.toRadixString(16)} (got 0x${got.toRadixString(16)})',
      got == clampedAt,
    );
    final otherMoved = report.unknown != 0 ||
        report.kindMismatch != 0 ||
        report.widened != 0 ||
        report.malformed ||
        report.refused != TableFixedRefusal.none;
    check(
      'C8 clamp $field raw=0x${rawValue.toRadixString(16)}: '
      'no other counter moved '
      '(unknown=${report.unknown} kindMismatch=${report.kindMismatch} '
      'widened=${report.widened} malformed=${report.malformed} '
      'refused=${report.refused})',
      !otherMoved,
    );
  }

  // The b32 and b64 cases cannot be FORGED PAST THE WIDTH CLAMP through
  // their declared storage (storage IS the clamp), so they are covered
  // by the round-trip case above — those two are ELIDED clamps.
  one('b8',  0x100,             _expectations['b8']!.maxValue);
  one('b12', 0x1000,            _expectations['b12']!.maxValue);
  one('b16', 0x10000,           _expectations['b16']!.maxValue);
  one('b48', 0x0001000000000000, _expectations['b48']!.maxValue);
}

void main() {
  print('C8: bits(N) storage width and width clamp');
  final layout = TableFixedLayout()
    ..at = 0
    ..view = ByteData.sublistView(rangedWidthsFixedLayout);
  final ok = layout.parse(
    ByteData.sublistView(rangedWidthsFixedLayout),
    0,
    rangedWidthsFixedLayout.length,
  );
  if (!ok) {
    print('FAIL C8 layout parse refused: ${layout.refusal}');
    exitCode = 1;
    return;
  }
  assertLayout(layout);
  assertRoundTrip();
  assertClamp();
  if (failures != 0) {
    print('C8 FAILED ($failures assertions)');
    exitCode = 1;
    return;
  }
  print('C8 green: bits(N) law asserted (kind, size, width clamp)');
}
